File: src/frontend/src/types/problem-set.ts
/**
 * 题单模块类型（docs/contracts/problem-sets.md）。
 * 团队题单可见性与团队题目对齐：team_visible（全队可见）/ admin_visible（仅团队管理）。
 */

/** 题单可见性：全站 public/private；团队 team_visible/admin_visible */
export type ProblemSetVisibilityType = 'public' | 'private' | 'team_visible' | 'admin_visible'

/** 题单内题目项（题目元信息随行返回） */
export interface ProblemSetItem {
  problem_id: string
  title: string
  difficulty?: number | null
  /** 题目限制（与题库列表口径一致） */
  time_limit_ms?: number
  memory_limit_mb?: number
  /** 当前用户作答状态（登录请求回填）：true=已通过 / false=已尝试未通过 / null|缺省=未提交过（未登录恒缺省） */
  solved?: boolean | null
  sort_order: number
}

export interface ProblemSetSummary {
  id: string
  title: string
  description?: string | null
  visibility: ProblemSetVisibilityType
  status: 'active' | 'archived'
  owner_id: string
  /** 归属团队（非空 = 团队题单；题单中心 / mine 恒为空，管理视图区分来源用） */
  team_id?: string | null
  item_count: number
  /** 非空 = 经团队引用进入团队题单（引用时间；团队空间） */
  referenced_at?: string | null
  created_at: string
  updated_at: string
}

/** 题单详情：题目按 sort_order 展示（刷题不强制按序完成） */
export interface ProblemSetDetail extends ProblemSetSummary {
  items: ProblemSetItem[]
  can_manage: boolean
  /** 创建人昵称（详情页展示；列表不携带） */
  owner_name: string
}

export interface ProblemSetCreatePayload {
  title: string
  description?: string | null
  visibility?: 'public' | 'private'
}

/** 编辑题单元信息（缺省不动，传即改） */
export interface ProblemSetEditPayload {
  title?: string
  description?: string | null
  visibility?: 'public' | 'private'
}

/** 编排题目：全量替换题单内列表；同题单内 problem_id 不得重复 */
export interface ProblemSetItemsPayload {
  items: Array<{ problem_id: string; sort_order: number }>
}

export interface ProblemSetListQuery {
  page?: number
  page_size?: number
  keyword?: string
  /** 题单中心「我的」勾选：仅本人未下线题单（含私有；须登录） */
  mine?: boolean
}

/** 管理视图查询（来源筛选：solo=全站题单 / team=团队题单） */
export interface AdminProblemSetListQuery extends ProblemSetListQuery {
  status?: 'active' | 'archived'
  ownership?: 'solo' | 'team'
}

File: src/backend/app/schemas/problem_set.py
"""题单模块请求 / 响应 Schema（docs/contracts/problem-sets.md）。"""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.enums import ProblemSetStatus, ProblemSetVisibility


class ProblemSetCreate(BaseModel):
    """创建题单：全站题单 public/private（admin/tutor）；团队题单随 teams 模块开放。"""

    title: str = Field(min_length=1, max_length=128)
    description: str | None = None
    visibility: ProblemSetVisibility = ProblemSetVisibility.PUBLIC

    @model_validator(mode="after")
    def check_visibility(self) -> ProblemSetCreate:
        if self.visibility == ProblemSetVisibility.TEAM:
            raise ValueError("团队题单随 teams 模块开放")
        return self


class ProblemSetUpdate(BaseModel):
    """编辑题单元信息（title / description / visibility 全量语义：传即改，缺省不动）。"""

    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = None
    visibility: ProblemSetVisibility | None = None

    @model_validator(mode="after")
    def check_visibility(self) -> ProblemSetUpdate:
        if self.visibility == ProblemSetVisibility.TEAM:
            raise ValueError("团队题单随 teams 模块开放")
        return self


class ProblemSetItemsUpdate(BaseModel):
    """编排题目：全量替换题单内列表；同一题单内 problem_id 不得重复。"""

    model_config = ConfigDict(extra="forbid")

    items: list["ProblemSetItemIn"] = Field(default_factory=list, max_length=200)


class ProblemSetItemIn(BaseModel):
    problem_id: uuid.UUID
    sort_order: int = Field(default=0, ge=0)


class ProblemSetItemOut(BaseModel):
    """题单内题目项（题目元信息随行返回，供详情页直接渲染）。"""

    model_config = ConfigDict(from_attributes=True)

    problem_id: uuid.UUID
    title: str
    difficulty: int | None = None
    sort_order: int


class ProblemSetSummary(BaseModel):
    """题单列表项（题单中心 / 管理视图共用）。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    description: str | None = None
    visibility: ProblemSetVisibility
    status: ProblemSetStatus
    owner_id: uuid.UUID
    item_count: int = 0
    created_at: datetime
    updated_at: datetime


class ProblemSetDetail(ProblemSetSummary):
    """题单详情：题目按 sort_order 展示（刷题不强制按序完成）。"""

    items: list[ProblemSetItemOut] = Field(default_factory=list)
    can_manage: bool = False


class ProblemSetSubmissionCreate(BaseModel):
    """题单内交题（POST /problem-sets/{id}/problems/{pid}/submissions）。

    problem_id 由路径提供（且必须属于该题单）；代码上限与练习提交一致（64KB UTF-8 字节）。
    """

    language: str = Field(max_length=32)
    code: str = Field(min_length=1, max_length=65536)

    @field_validator("code")
    @classmethod
    def code_bytes_limit(cls, value: str) -> str:
        if len(value.encode("utf-8")) > 64 * 1024:
            raise ValueError("代码不能超过 64KB")
        return value
